### ISSUE TITLE
session lifetime?

### ISSUE THREAD

**Issue by ibrokhimjalolov** [Kirill-Lekhov/ya-business-api#6]
Is there any information about session lifetime?

**Comment by Kirill-Lekhov** [comment]
Hello!

According to my personal observations, the lifetime of session is only a few days (1-3).

**Comment by kusainovv** [comment]
## Great topic! 
### Please don't close this issue until it's resolved

I wrote about it [before](https://github.com/Kirill-Lekhov/ya-business-api/issues/4) and got the answer but this doesn't solve the issue.

#### I have thoughts about it: 

1. Develop the service for getting the latest reviews by each hour. This ensure that we keep the session active and emulate user activity. 

2. Store the newest value from each request and update the ```reviews_csrf_token``` and ```answer_csrf_token``` tokens by getting them from cookies (but I see these cookies are httpOnly and secure)

3. Develop the service using a headless browser (such as Playwright or Selenium) and run the ```getData()``` function (**code below**) but this **increases the difficulty** because I tried to implement service like this using JavaScript and Puppeteer, and this way was impossible because the Yandex System identified the request as wrong and display a CAPTCHA.

```js
function getData() {
	console.info({
		"CSRFToken": window?.__PRELOAD_DATA?.initialState?.env?.csrf,
		"PermanentId": window?.__PRELOAD_DATA?.initialState?.edit?.company?.permanent_id,
	})
}


getData()

/**
 * {CSRFToken: "...", PermanentId: 00000000000}
*/
```

## In summary
The best way to keep credentials available is to emulate user interactivity, but this doesn't apply to ```csrf_token``` because obtaining the ```csrf_token``` is only possible on the client-side of an authenticated user. Emulating an authenticated user is difficult to implement because the Yandex Service identifies the request as suspicious and displays the CAPTCHA.

Any thoughts? Maybe I can help you?

**Comment by Kirill-Lekhov** [comment]
Have you tried not using the headless mode of the browser driver? How does the site behave in this case?

**Comment by kusainovv** [comment]
I'm about to translate my thoughts into code, so I will show you the example as soon I develop it, please wait!

**Comment by Kirill-Lekhov** [comment]
I spent some time developing a prototype that is capable of updating the main authorization cookie.

```python3
from sys import exit

import undetected_chromedriver as uc


def main() -> int:
	options = uc.ChromeOptions()
	options.add_argument("--disable-dev-shm-usage")
	options.add_argument("--disable-infobars")
	options.add_argument("--disable-extensions")
	options.add_argument("--no-sandbox")
	driver = uc.Chrome(headless=False, options=options, driver_executable_path="/usr/bin/chromedriver")

	driver.execute_cdp_cmd(
		"Network.setCookie",
		{
			"name": "Session_id",
			"value": input("Please, enter Session_id cookie value: "),
			"domain": ".yandex.ru",
			"path": "/",
			"sameSite": "None",
			"secure": True,
		}
	)
	driver.get("https://yandex.ru/sprav/")
	print(driver.get_cookie("Session_id"))

	return 0


if __name__ == "__main__":
	exit(main())
```

This script reads the old Session_id and outputs the new one to the console. So you can write a script or an authorization server using this solution. 

For docker container i recommend to use [pyvirtualdisplay](https://github.com/ponty/PyVirtualDisplay).

```python3
from sys import exit

import undetected_chromedriver as uc
from pyvirtualdisplay.display import Display


def main() -> int:
	options = uc.ChromeOptions()
	options.add_argument("--disable-dev-shm-usage")
	options.add_argument("--disable-infobars")
	options.add_argument("--disable-extensions")
	options.add_argument("--no-sandbox")

	with Display(visible=False):
		driver = uc.Chrome(headless=False, options=options, driver_executable_path="/usr/bin/chromedriver")

		driver.execute_cdp_cmd(
			"Network.setCookie",
			{
				"name": "Session_id",
				"value": input("Please, enter Session_id cookie value: "),
				"domain": ".yandex.ru",
				"path": "/",
				"sameSite": "None",
				"secure": True,
			}
		)
		driver.get("https://yandex.ru/sprav/")
		print(driver.get_cookie("Session_id"))

	return 0


if __name__ == "__main__":
	exit(main())
```

**Comment by kusainovv** [comment]
## Ops, CAPTCHA!

This is a code that I run:

```python

load_dotenv()

app = FastAPI()

session_id = os.getenv('YANDEX_BUSINESS_SESSION_ID')
session_id2 = os.getenv('YANDEX_BUSINESS_SESSION_ID2')
csrf_token = os.getenv('YANDEX_BUSINESS_CSRF_TOKEN')

api = SyncAPI.build(
    session_id=session_id,
    session_id2=session_id2,
    csrf_token=csrf_token,
)

def main() -> int:
    options = uc.ChromeOptions()
    options.add_argument("--disable-dev-shm-usage")
    options.add_argument("--disable-infobars")
    options.add_argument("--disable-extensions")
    options.add_argument("--no-sandbox")

    driver = uc.Chrome(headless=False, options=options)  # driver_executable_path="/path/to/chromedriver"

    driver.execute_cdp_cmd(
        "Network.setCookie",
        {
            "name": "Session_id",
            "value": "", # at my code, here's a value
            "domain": ".yandex.ru",
            "path": "/",
            "sameSite": "None",
            "secure": True,
        }
    )
    driver.execute_cdp_cmd(
        "Network.setCookie",
        {
            "name": "sessionid2",
            "value": "", # at my code, here's a value
            "domain": ".yandex.ru",
            "path": "/",
            "sameSite": "None",
            "secure": True,
        }
    )

    driver.get("https://yandex.ru/sprav/")
    print(driver.get_cookie('Session_id'), driver.get_cookie('sessionid2'))

    return 0

main()
```

This is an output:


![Screenshot](https://github.com/user-attachments/assets/b99de37a-bd6d-459f-8f6b-4920dd7a22eb)


So I'm going to use a back-proxy or python libraries that can ensure the request isn't suspicious

**Comment by Kirill-Lekhov** [comment]
@kusainovv

Can you give a complete example of the code that you run on your machine? I see discrepancies in the example code with the code that you have in the IDE.

Also, first try to perform a simple token update (with using my example), without initializing the FastAPI server.

Also, if you open the browser manually and see a captcha on this page, your IP address may be blacklisted.

**Comment by kusainovv** [comment]
You're right, your code works as expected, and I get the result without CAPTCHA.

<hr />

> Also, if you open the browser manually and see a captcha on this page, your IP address may be blacklisted.

Yep, thank you for the note because I encountered it before!

At the moment, the code works as expected, thank you for this, I'm gonna test it again and run this in Docker.

I'll write about it if I can do it!

**Comment by kusainovv** [comment]
> I spent some time developing a prototype that is capable of updating the main authorization cookie.
> 
> ```python
> from sys import exit
> 
> import undetected_chromedriver as uc
> 
> 
> def main() -> int:
> 	options = uc.ChromeOptions()
> 	options.add_argument("--disable-dev-shm-usage")
> 	options.add_argument("--disable-infobars")
> 	options.add_argument("--disable-extensions")
> 	options.add_argument("--no-sandbox")
> 	driver = uc.Chrome(headless=False, options=options, driver_executable_path="/usr/bin/chromedriver")
> 
> 	driver.execute_cdp_cmd(
> 		"Network.setCookie",
> 		{
> 			"name": "Session_id",
> 			"value": input("Please, enter Session_id cookie value: "),
> 			"domain": ".yandex.ru",
> 			"path": "/",
> 			"sameSite": "None",
> 			"secure": True,
> 		}
> 	)
> 	driver.get("https://yandex.ru/sprav/")
> 	print(driver.get_cookie("Session_id"))
> 
> 	return 0
> 
> 
> if __name__ == "__main__":
> 	exit(main())
> ```
> 
> This script reads the old Session_id and outputs the new one to the console. So you can write a script or an authorization server using this solution.
> 
> For docker container i recommend to use [pyvirtualdisplay](https://github.com/ponty/PyVirtualDisplay).
> 
> ```python
> from sys import exit
> 
> import undetected_chromedriver as uc
> from pyvirtualdisplay.display import Display
> 
> 
> def main() -> int:
> 	options = uc.ChromeOptions()
> 	options.add_argument("--disable-dev-shm-usage")
> 	options.add_argument("--disable-infobars")
> 	options.add_argument("--disable-extensions")
> 	options.add_argument("--no-sandbox")
> 
> 	with Display(visible=False):
> 		driver = uc.Chrome(headless=False, options=options, driver_executable_path="/usr/bin/chromedriver")
> 
> 		driver.execute_cdp_cmd(
> 			"Network.setCookie",
> 			{
> 				"name": "Session_id",
> 				"value": input("Please, enter Session_id cookie value: "),
> 				"domain": ".yandex.ru",
> 				"path": "/",
> 				"sameSite": "None",
> 				"secure": True,
> 			}
> 		)
> 		driver.get("https://yandex.ru/sprav/")
> 		print(driver.get_cookie("Session_id"))
> 
> 	return 0
> 
> 
> if __name__ == "__main__":
> 	exit(main())
> ```

<hr />

I'm using the Windows OS and the pyvirtualdisplay makes unexpected problems because pyvirtualdisplay uses the Xvfb and the Windows doesn't support this.

So, could you please share the Dockerfile?

**Comment by kusainovv** [comment]
## I created the Dockerfile

```dockerfile
FROM ubuntu:22.04
FROM python:3.9-slim

ENV PYTHONUNBUFFERED=1

# Update package lists for the Ubuntu system
RUN apt-get update && \
    apt-get install -y \
        wget \
        unzip \
        xvfb \
        curl
RUN pip install selenium==4.0.0

# Install the 'unzip' package
RUN apt install unzip -y

# Copy the Chrome Debian package to the image
COPY chrome_114_amd64.deb ./

# Install the Chrome Debian package
RUN apt install ./chrome_114_amd64.deb xvfb -y

# Download ChromeDriver binary version 114.0.5735.90 for Linux
RUN wget https://chromedriver.storage.googleapis.com/114.0.5735.90/chromedriver_linux64.zip


# Unzip the downloaded ChromeDriver binary
RUN unzip chromedriver_linux64.zip

# Move the ChromeDriver binary to /usr/bin
RUN mv chromedriver /usr/bin/chromedriver

# Download uBlock Origin extension
RUN wget -O ublock_origin.crx "https://clients2.google.com/service/update2/crx?response=redirect&prodversion=91.0.4472.101&x=id%3Dcjpalhdlnbpafiamejdnhcphjbkeiagm%26installsource%3Dondemand%26uc"

# Print the version of Google Chrome installed
RUN google-chrome --version



WORKDIR /app

COPY requirements.txt .
RUN pip install -r requirements.txt

COPY . .

CMD ["uvicorn", "main:app", "--host", "0.0.0.0", "--reload"]
```


<hr />


### However, your code anyway gets the CAPTCHA. 

I try to use back proxy (Bright Data) for bypassing the captcha but this doesn't work. Now, I'm trying to find the bypassing service for money,

**Comment by kusainovv** [comment]
So I don't have any thoughts about it.

I have the cron function for updating the authorization credentials and this doesn't work.

```typescript
def emulate_user_interactivity_callback():
    # Get a fresh CSRF token
    fresh_csrf_token = api.service.get_csrf_token()
    # Call the user interactivity function with the fresh CSRF token
    run_emulate_user_interactivity(csrf_token=fresh_csrf_token)

scheduler = BackgroundScheduler()
scheduler.start()
scheduler.add_job(emulate_user_interactivity_callback, 'cron', hour='8,20', minute=0, max_instances=1)

```

```typescript

async def run_interactivity(csrf_token: str):
    auto_scrape_credentials(csrf_token, True)
    return 1

def run_emulate_user_interactivity(csrf_token: str):
    asyncio.run(run_interactivity(csrf_token))
    
```


```typescript

from dotenv import load_dotenv
import os

load_dotenv()

class CookieStore:
    def __init__(self):
        self.cookies = {
            'Session_id': os.getenv('YANDEX_BUSINESS_SESSION_ID'),
            'sessionid2': os.getenv('YANDEX_BUSINESS_SESSION_ID2'),
            'csrf_token': os.getenv('YANDEX_BUSINESS_CSRF_TOKEN'),
        }

    def set_cookie(self, key: str, value):
        self.cookies[key] = value  

    def get_cookie(self, key: str):
        return self.cookies.get(key)
    
    def get_all_cookies(self):
        return self.cookies  

cookie_store = CookieStore()

```

```typescript
import undetected_chromedriver as uc
import time
import os
from state import cookie_store
from pyvirtualdisplay.display import Display
import random
from fake_useragent import UserAgent

session_id = os.getenv('YANDEX_BUSINESS_SESSION_ID')
session_id2 = os.getenv('YANDEX_BUSINESS_SESSION_ID2')

ua = UserAgent()

def get_random_user_agent():
   return ua.random

def auto_scrape_credentials(csrf_token: str):
   options = uc.ChromeOptions()
   options.add_argument("--disable-dev-shm-usage")
   options.add_argument("--disable-infobars")
   options.add_argument("--disable-extensions")
   options.add_argument("--no-sandbox")
   
   options.add_argument(f"user-agent={get_random_user_agent()}")
   options.add_argument(f"referer=https://passport.yandex.ru/auth?retpath=https://yandex.ru/sprav/companies")
   
   driverPath = '/usr/bin/chromedriver'
   
   with Display(visible=False):
      driver = uc.Chrome(headless=False, driver_executable_path=driverPath , version_main=114, options=options)

      driver.execute_cdp_cmd(
               "Network.setCookie",
               {
                  "name": "Session_id",
                  "value": session_id,
                  "domain": ".yandex.ru",
                  "path": "/",
                  "sameSite": "None",
                  "secure": True,
               }
         )
         
         
      driver.execute_cdp_cmd(
               "Network.setCookie",
               {
                  "name": "sessionid2",
                  "value": session_id2,
                  "domain": ".yandex.ru",
                  "path": "/",
                  "sameSite": "None",
                  "secure": True,
               }
         )
         
      driver.get("https://yandex.ru/sprav")
      
      time.sleep(random.uniform(2, 7))
                  
      html_content = driver.page_source
         
      if "captcha" in html_content:
         print(" Ops, CAPTCHA! ")
               
      data = driver.execute_script("""
            return {
               "CSRFToken": window?.__PRELOAD_DATA?.initialState?.env?.csrf,
            };
      """)
            
      cookie_store.set_cookie('Session_id', driver.get_cookie('Session_id')['value'])
      cookie_store.set_cookie('sessionid2', driver.get_cookie('sessionid2')['value'])
      cookie_store.set_cookie('csrf_token', data if data.get('CSRFToken') != None else csrf_token)

      print(cookie_store.get_all_cookies())
    
      driver.quit()
   
   return 0
```


This web scraping for emulating user interactivity doesn't work the reasion is SmartCaptcha by Yandex. 

Do you have any ideas about it 😀 ?

**Comment by Kirill-Lekhov** [comment]
I don't think this topic should continue because there is a solution to extend authorization and it has been tested.

I suggest that we continue discussing the problem of bypassing the captcha by email. Please send me the full code of your project (without credentials) by e-mail kirill.lekhov@mail.ru

**Comment by kusainovv** [comment]
Yep, I agree, it's clear
